### DISCUSSION TITLE
FreeBSD Support

### DISCUSSION THREAD

**Discussion by FalloutZone** [microsoft/WSL#6540] [General]
Support FreeBSD Binaries and System Calls

Expected Result:
Run FreeBSD on the subsystem

FreeBSD is a very popular server operating system with deep roots in web and cloud development.  Please support FreeBSD in the Windows 10 "WSL" feature.

**Comment by drsn0w** [comment]
This is Windows Subsystem for _Linux_, not FreeBSD or other *NIX environments.

**Comment by FalloutZone** [comment]
Yes, you are correct - WSL is for ELF/Linux.  This is a request to make this, or a similar system on Windows 10 WSL to support other open source *NIX environments - specifically FreeBSD.

**Reply by kilitary** [reply]
they even can make frozen versions of lib and loading elf or a.out will be just a user ring asm code.
but in reality i do not clearly know how WSL is implemented because i do not need this. wsl is good.

**Comment by sunilmut** [comment]
@FalloutZone - Thanks for the feedback. What you are essentially asking for is "WSB" (Windows Subsystem for BSD). [User voice ticket](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/category/161892-bash) would be a great way to provide that feedback. Something that we look at for our future prioritization.

**Reply by maphew** [reply]
"This UserVoice instance is no longer available."

**Comment by eeboaz** [comment]
I would like to see this as well. I prefer the FreeBSD operating system over Linux for server and development use cases and this would make Windows 10 a much more viable option for a development platform for me. I will look at the User voice ticket link provided.

**Comment by iz0eyj** [comment]
it may be that it becomes useful in a possible version for use on servers, but the use of *BSD in desktop machines is really very limited.

**Comment by Onepamopa** [comment]
I would also like to see Windows Subsystem for FreeBSD.

**Comment by eeboaz** [comment]
> it may be that it becomes useful in a possible version for use on servers, but the use of *BSD in desktop machines is really very limited.

There's no good reason to arbitrarily permit *BSD only on servers but not on desktops.
The use case for WSL is not the server market - it's the developer desktop market.

BSD is much the same. This would let me have *one* system at work where I have to use Windows for corporate stuff (exchange/email/calendar, visio, etc.) and BSD where I code for servers running FreeBSD
Apple is continuing to make a mockery of macOS and isn't really viable any more.

**Comment by iz0eyj** [comment]
@eeboaz:
I have absolutely nothing against BSD, but his primary field of use are the servers and network devices. The WSL we have now is not suitable for use on any of the two types of object, and the effort needed to create in "WSB", at least in my opinion, it would be totally unjustified.
If later the characteristics of WSL will change in a direction that will be suitable for those uses, then it will change my opinion.

**Comment by eeboaz** [comment]
@iz0eyj 
Why are you insisting on this view point that we're wanting to run production systems on WSL/WSB?

This is exactly counter to the purposes of WSL in the first place, straight from the [FAQ](https://msdn.microsoft.com/en-us/commandline/wsl/faq):

> Who is this for?
>
> This is primarily a tool for developers -- especially web developers and those who work on or with open source projects. This allows those who want/need to use Bash, common Linux tools (sed, awk, etc.) and many Linux-first tools (Ruby, Python, etc.) to use their toolchain on Windows.

**Comment by iz0eyj** [comment]
@eeboaz:
where you see reported the BSD say in what you brought? if anything at all WSL documents contain references to Linux and exclusively to Linux. For a project to be successful it needs there are users who then use it ... the development team has a cost, are not eople who work for the glory.
What do you prove by that passage? which it is directed ONLY to the developers? is not the case, this is what he would like some of them but it does not. For any developer who receive WSL exist thousand and thousand ordinary users; It will be sufficient for an ordinary user of a hundred active WSL to see them dominate much the number of developers.
There are no end-user Windows interested in BSD, they want Ubuntu Linux.
WSL is not a product for the community "free software" but it is a Windows component, whose main target will be Windows users, be they end users or developers.

**Comment by eeboaz** [comment]
@iz0eyj 
> There are no end-user Windows interested in BSD, they want Ubuntu Linux.

I did't know you spoke for the entire Windows community. Right, you don't. 
You're not going to dissuade me from being vocal that I would like this support.

We're just asking the development team at Microsoft to consider supporting FreeBSD in addition to Linux when they've reached their primary goal of supporting Ubuntu Linux. I don't see how having options harms you at all, so please be considerate and let me and others ask for the same option: using the Windows subsystem to support the user-land of the operating system we prefer to use for our development work.

**Comment by rustyx** [comment]
I'm surprised this "issue" is still open - this project is about Linux emulation, after all.
BSD user mode is _way_ more tightly coupled to the kernel than Linux. Given that, plus the dynamic nature of the FreeBSD ecosystem, emulating a BSD kernel would be an exercise in futility.
Which is no problem at all because there is very, very few reasons to emulate BSD anyway.
You can never use it for testing - if your code doesn't work, you'd _always_ want to try it on a real BSD.

It's not about having the option, it's about spending hundreds or thousands of development hours to create and maintain something that is marginally useful.

**Comment by aseering** [comment]
G'day all -- there are some valuable opinions here, but could I request that we keep this thread on topic?

The topic here is a request for support for a BSD userspace.  That is a valid technical request.  It is up to the WSL team to pick it up or not.  I, for one, trust them to have sound judgment.  If you have supporting use cases or technical commentary on making that happen, please post it here.

If, on the other hand, you think this is a totally silly idea, stop replying and bumping up the ticket's post-count :-)  According to this repository's [README](https://github.com/Microsoft/BashOnWindows/blob/master/README.md), issues remain open until the "discussion has run its course."  So I encourage you to resist the urge to correct the thing that you see as Wrong On The Internet, and let the thread come to a stop.

**Comment by therealkenc** [comment]
User Voice is [here](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/31866838-support-bsd-operating-systems).

**Comment by ruffianeo** [comment]
The one reason I would love to see FreeBSD support (even instead of linux) is that I have a deadly GPL allergy and I try to stay clear of such (legal) issues by avoiding Linux whenever I can. The "mostly for network and servers" argument does not cut it IMHO. FreeBSD is THE Unix for people who hate GPL.

**Reply by ghost** [reply]
Same

**Comment by Biswa96** [comment]
As WSL2 now relies on virtual machine, if BSD has a proper Hyper-V socket support it may be possible to port the init binary to BSD. The rest is just socket read-write and pty.

**Comment by ghost** [comment]
> 
> 
> As WSL2 now relies on virtual machine, if BSD has a proper Hyper-V socket support it may be possible to port the init binary to BSD. The rest is just socket read-write and pty.

Looks like there's a WIP: https://reviews.freebsd.org/D24061

**Comment by ykla** [comment]
Also hope Windows will supports FreeBSD.

**Comment by xpader** [comment]
Microsoft love Linux, not FreeBSD..
They just stolen code from FreeBSD~~
Maybe WSL should be rename to WSU (Windows Subsystem for Unix).

**Reply by ghost** [reply]
It wasn't stealing tho. BSD people are actually generous; they made their software truly free as in freedom

**Comment by connstructor** [comment]
> > As WSL2 now relies on virtual machine, if BSD has a proper Hyper-V socket support it may be possible to port the init binary to BSD. The rest is just socket read-write and pty.
> 
> Looks like there's a WIP: https://reviews.freebsd.org/D24061

This is a dead link. I, too, would love to see a "WSB" capability. I can run BSD just fine in Hyper-V.

It absolutely stuns me how much ignorance and misinformation I'm seeing in this issue, by the way.

> the use of *BSD in desktop machines is really very limited.

Maybe you don't, but I certainly do. I have three workstations in my office: Windows for gaming (I'm lazy, sorry), Linux, and FreeBSD. Anything I TRULY care about, I do on the FreeBSD machine, because I require absolute reliability in my work.

> primary field of use are the servers and network devices.

See above.

> For a project to be successful it needs there are users who then use it

No one uses FreeBSD? News to me.

This one is particularly egregious:

> There are no end-user Windows interested in BSD, they want Ubuntu Linux.

I'm a Windows user, as stated previously. I also use Linux, and I also use BSD. I also won't touch Ubuntu with a six-parsec long pole. And yes, parsecs are a unit of distance, not time.

> there is very, very few reasons to emulate BSD anyway.

Aside from everything a BSD user would do on a desktop or for development?

> You can never use it for testing

What? I don't even have a pithy response for this one.

I would encourage anyone wanting to further comment on this to actually familiarize yourself with what *BSD is instead of spouting stereotypes and complete nonsense here.

**Comment by jmb49** [comment]
> Looks like there's a WIP: https://reviews.freebsd.org/D24061
> This is a dead link.

Access granted today from France (2021, Fev 06th). FYI.

**Comment by jmb49** [comment]
> Maybe WSL should be rename to WSU (Windows Subsystem for Unix).

Before WSL Redmond's Seven brought us Win SUA. I let you discover what it was ; but it was close and it was not sLa...

**Comment by connstructor** [comment]
> The link works in my case.

Yeah, works for me now, too. Oops? 😛

**Comment by jmb49** [comment]
> If init and initrd were free and open source WSB would be a real thing now.

I think this is the point OR... well I must amdmit I could have misunderstood everything :)
PS : absolutely no pb with you, f0x1de, especialy Star Treck fanatic :+1:

**Comment by ykla** [comment]
This UserVoice instance is no longer available. Please reopen this issue.

**Comment by kevincrans** [comment]
What about Chrome OS support, or steamos? The only thing bsd and linux have in common is it's unix-like monolithic kernels.

**Comment by laffer1** [comment]
Hyper-v socket support landed in FreeBSD current at least. 
https://github.com/freebsd/freebsd-src/commit/a560f3ebd77733208fa8371a5f2d09523e847c0d

**Comment by tony** [comment]
To participants that thumbs upped this, please upvote the [original post](https://github.com/microsoft/WSL/discussions/6540#discussion-3216707) as this has been moved to a discussion.

**Comment by umop3plsdn** [comment]
does anyone know if this is being bounced around or if it's a dead cause?

**Comment by tony** [comment]
Not sure how many users it has - and this may be a chicken or egg thing, but Azure supports FreeBSD:

https://docs.microsoft.com/en-us/azure/virtual-machines/linux/freebsd-intro-on-azure

(On the other hand, another large cloud service that supported FreeBSD, recently [demoted their support to a custom image](https://www.digitalocean.com/blog/latest-products-and-features-april-2022))

If @microsoft were to sponsor FreeBSD and do a WSL-like solution for it, there'd be a permissively-licensed subsystem in Windows. But how? HyperV? Is there a public API to make your own Windows Subsystem(s)?

I don't know what it'd look like, but FreeBSD's focus on developer quality of life almost invites this.

That said - despite FreeBSD having big adopters and a strong userbase and great, skilled maintainers - it needs quite a few strokes of luck, or a confluence of favorable circumstances, to gain momentum to be a big player in the cloud / gain critical mass.

P.S. I'm not sure how other BSDs weigh in in terms of compatibility, but they're similarly developer focused.

**Reply by xpader** [reply]
WSL is Windows Subsystem for Linux, not Windows Subsystem for Unix..

**Reply by tony** [reply]
- https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/feature-descriptions-for-linux-and-freebsd-virtual-machines-on-hyper-v
- https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/supported-freebsd-virtual-machines-on-hyper-v
- https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/best-practices-for-running-freebsd-on-hyper-v

**Reply by ruffianeo** [reply]
The topic is about a WSFBSD :) I am not sure how that is related to anything cloud.

**Comment by sydbarrett74** [comment]
+1 /bump

Since the BSD licence isn't viral (to use Steve Ballmer's turn of phrase), I should think Microsoft would be even _more_ enthusiastic to incorporate it into the Windows ecosystem. There are lots of use-cases for FreeBSD on a Windows box.

**Comment by ghost** [comment]
Would you be more interested in implementing this if I reiterated our request in a new issue/discussion with cute cat photos?
Would you like a song? "I'm beggin', beggin' you, so put your loving hands on this project". Bribes perhaps? Idk; Anything! Just do it, please.

**Comment by ykla** [comment]
The current wsl2 implementation is entirely based on the hyperV VM, so I don't think there should be any technical difficulties with it. Microsoft should consider supporting FreeBSD.

**Comment by ghost** [comment]
I'm gonna post on a weekly-ish basis till someone pays attention to this thread

**Comment by ghost** [comment]
![image](https://user-images.githubusercontent.com/20170702/216603060-5e36c472-7f57-4a23-90f1-5f5b7d99248d.png)

**Comment by ghost** [comment]
WSBSD may actually incentivise me to suck up driver issues and upgrade my other device from 7 to 10.

**Comment by ghost** [comment]
Hello there

**Comment by nttranbao** [comment]
+1 for me. It would be great to have FreeBSD on WSL/WSBSD

**Comment by yesudeep** [comment]
+1 for extra effect.

**Comment by sirredbeard** [comment]
One idea, which [I have experimented with out of curiosity](https://twitter.com/unixterminal/status/1291960047791153154?s=20), is to build an incredibly minimal Linux image, using LFS, but then populate it with packages from pkgsrc. I realize this is not a BSD kernel, but it provides much of the same BSD-like userland experience, including package management, with the performance optimizations/integrations built into the WSL Linux kernel.

**Comment by ghost** [comment]
@sirredbeard Really cool but still... not BSD

**Comment by ghost** [comment]
One does wonder though that if Windows were to have a subsystem for BSD; Would the daemons still make their signature BSD "scratchy sound" or would Windows shut them up in a virtual void, offering a less daunting, user friendly experience. I admit that it does make me worry a bit when I run BSD on metal.

(https://docs.freebsd.org/en/books/faq/#letmeoutofhere)

**Comment by ghost** [comment]
Sorry for the delay; I had covid

**Comment by ghost** [comment]
I would've announced that I can finally breathe properly with a joke but I can't think of any

**Comment by ghost** [comment]
BSDs being actually free as in freedom could've meant that Microsoft could just copy a lot of kernel code directly and just implemented a sort of shim between BSD interface and NT kernel. Maybe even contributing a proper NTFS driver back to upstream.
It still isn't late though! WSBSD can still be what WSL 1 could've never been.

**Comment by osen** [comment]
Microsoft used to have Services for UNIX (SFU): https://en.wikipedia.org/wiki/Windows_Services_for_UNIX

This was in part based on BSD (OpenBSD) for the userland as a remnant of MKS Toolkit.

This (sadly) did not see the same popularity as the WSL hype. I can understand why Microsoft won't invest here.

**Reply by ghost** [reply]
I believe SFU was actually ports of Unix software using a compatibility library like Cygwin or libbsd for Linux. You couldn't actually run any old OpenBSD ELF using SFU. WSL however runs Linux binaries just fine.

**Reply by osen** [reply]
(Edit: Sorry for so many edits, the history of SFU, SUA is messy!)

SFU was more like WSL1 rather than the more virtualization / Hyper-V based approach of WSL2.
But it was actually a subsystem, Obviously unlike something like Cygwin, the executables built for SFU couldn't be made standalone to run on Windows without the subsystem.

> You couldn't actually run any old OpenBSD ELF using SFU. WSL however runs Linux binaries just fine.

SFU needed to be thought of as a distinct UNIX platform. So no it couldn't run unchanged OpenBSD binaries any more than Solaris could. However I believe it did run unchanged Interix binaries as part of its Interix Subsystem.

It was all there, we are just going in circles these days ;)

**Reply by ghost** [reply]
Perhaps if Microsoft did a better job of informing people like me what SFU was it wouldn't have been such a flop.
Still... I'd love WSBSD

**Reply by kevincrans** [reply]
It's available on windows 7 based os. I think both ms and apple could do a better job as for their own os. However, it's hard since every bsd and version number is different.

**Comment by ghost** [comment]
osen's post and my comments kept the thread active; Seems like my weekly cry for attention to this thread isn't needed at this time....
Oops I did just that I guess

**Comment by LegalizeAdulthood** [comment]
I'll add that I'd also like to see BSD support in a supported Windows Subsystem type product.

**Comment by ghost** [comment]
![image](https://user-images.githubusercontent.com/20170702/234909353-4dff7458-a6d1-466b-85fa-53915fa6bc66.png)
gimme ma bsd!

**Comment by starXultra** [comment]
I honestly want windows subsystem for BSD because of the **userland**, couldn't care less for the kernel.
**The kernel isn't going to make much difference here, because the BSD tools you use are the userland, and the structure and benefits of the BSD kernel aren't going to have much benefit because you are running it on top of the NT kernel (The kernel that is used by windows)**
Trying to port **BSD userland to Linux and putting it in WSL could work,** as **this has been done in a native Linux distribution called Chimera Linux**
Maybe we can get the source code and rip the userland from either the Chimera Linux code or FreeBSD.
Source: [The Chimera Linux official website](https://chimera-linux.org/#:~:text=its%20userland%20tools%20are%20based%20on%20FreeBSD)

**Reply by kevincrans** [reply]
That's a good starting point! however people like their own bsd. I read that crux linux also has bsd style init.

**Reply by starXultra** [reply]
I don't think init will matter because wsl uses its own init
I think WSL just uses a hacky program pretending to be a linux kernel so that binaries can run, so kernel still doesn't matter. Linux is a clone of unix and BSD is unix, so it doesn't matter.
I'm going to try and bootstrap chimera and chroot into it in WSL, as I have downloaded chroot tarballs and used them in WSL before, and they worked with no problems.

**Reply by kevincrans** [reply]
Interesting, but my point still stands. BSD is not officially unix (macos aside), but that may have worked in Chimera Linux' advantage.

**Reply by starXultra** [reply]
BSD is a descendent of unix, but they slowly worked on removing at and t code. It works exactly like unix, even though it has no unix code, unlike linux, which just pretends to work like unix.

**Reply by ghost** [reply]
Well not _exactly_ like Research Unix but I'd argue that they've done a far better job in improving on it (kqueue, jails, etc)

**Reply by unrooted** [reply]
hi, I've just found this thread
if anyone is interested in using Chimera, I've tried their rootfs minimal .tar.gz and seems like it's working perfectly fine, dunno about GUI support for it whether it requires any further tweaks, but it seems to work perfectly fine as terminal-only (but as you can see I've just added repos and not much beyond that so far)
![image](https://github.com/microsoft/WSL/assets/30440603/27301f5a-0e10-4b6f-b37a-5b971e7ca616)

**Comment by ghost** [comment]
![image](https://github.com/microsoft/WSL/assets/20170702/934cc28e-32f5-40d3-9238-4e52aadc22cc)

**Reply by kilitary** [reply]
like a char m

**Comment by ghost** [comment]
knock knock

**Reply by kilitary** [reply]
обьект обнаружен ... обработка ... но нам нахуй он не нужен

**Comment by nikolaizamega** [comment]
How to patch KDE2 for FreeBSD? 😂

**Comment by ris-work** [comment]
I'd love to see any BSD support as well. NetBSD already runs as a RUMPkernel (and has been compiled to JS as well!) so I suppose it should not be very different to what's done for Linux. Moreover, I suppose the license is far more lenient, too.
Since it already uses a hypervisor anyway, it should at least be bootable?

**Comment by netjune** [comment]
I think it should be easy to support FreeBSD using hyper-V like WSL2.

**Comment by trungnt2910** [comment]
For the NT kernel driver devs out there:

As of Windows 11 23H2, it is [possible](https://github.com/microsoft/WSL/discussions/4022#discussioncomment-7717130) to create custom Pico providers/library OSes on Windows.

Currently I have created a [toy subsystem](https://github.com/trungnt2910/lxmonika/tree/master/mxss) that can print "Hello World" using a custom syscall ABI. In the mid-term future, I might attempt OSes of mid-complexity like [Haiku OS](https://github.com/haiku/haiku), before proceeding to newer Linux/Darwin/FreeBSD in the long run.

I'm current wading through the NT kernel internals and learning new things every day, [join me](https://reality.trungnt2910.com/discord/lxmonika) if you can contribute or just wanna have a chat.

**Comment by ghost** [comment]
![image](https://github.com/microsoft/WSL/assets/20170702/e13b2c9f-43a0-4ddf-8513-ef1aa993b63f)

**Comment by Nolram12345** [comment]
+1 on this. BSD for WSL (WSU?) would be awesome.

**Reply by xzenor** [reply]
+1 on this. Not that I expect it will ever happen but still... Worth a try

**Comment by DeviNoles** [comment]
+1

**Comment by pleasesendmeloadsofvegetables** [comment]
Can we get an official stance on this? The last time a maintainer posted in this discussion was in 2017. I would love to use WSBSD (or whatever it would be called) instead of WSL.

**Comment by ManaHime** [comment]
With wsl becoming mostly open source, I wonder if this means that wsbsd might become possible? ＾＾

**Reply by trungnt2910** [reply]
It should be possible.

With all the interop code open-source, we can theoretically start a distro that is a thin wrapper around KVM, then use some magic sockets to forward the interop stuff from Windows to the FreeBSD guest.

**Reply by xzenor** [reply]
It becomes more efficient to just run it in Hyper-V then...

**Reply by trungnt2910** [reply]
The WSL2 architecture has never been about efficiency.

**Reply by VOM7HC** [reply]
Thank @trungnt2910, I also think WSL 1 is better performance than WSL 2 (which based on Hyper-V), but one will stuck that Windows is never be POSIX compliant so if you have app or program call POSIX syscall API like `fork()` this will meet unexpected behavior. How your handle it? I still see micro kernel concept like WSL 2 is suitable for POSIX compliant if run binary without modify

I saw all your comment and posts related with "Revive Astoria", LXMonika and other comment relate WSL discussion here. I am also interest all of these. Because I very interested in Windows Programming. But I have different view of this path, like WSX (Windows Subsystem for XNU) which will run some app like xcode like darling, which will contain some code from WinObjC and [ipasim](https://github.com/ipasimulator/ipasim).

Are you Vietnamese? Can we share contact together?

**Reply by trungnt2910** [reply]
@VOM7HC 

> but one will stuck that Windows is never be POSIX compliant

Windows used to be POSIX-compliant with its own POSIX subsystem. POSIX itself is not a big issue, since it is a rather high-level standard, leaving a lot of room for implementation-specific behavior.

In fact, WSL1 handles plain POSIX syscalls quite well, since they're well maintained in WSL1's prime in 2016 - 2017. It is only when Linux-specific syscalls/features (e.g. `epoll`, `/proc` files) get involved will things fall apart.

> How your handle it? I still see micro kernel concept like WSL 2 is suitable for POSIX compliant if run binary without modify

WSL2 is not a micro-kernel in its pure sense. It is not a component dependent on Windows, it is an entirely separate virtual machine (just like VirtualBox) with some magic sockets.

I do agree that to get full _Linux_ compatibility (not POSIX), we need to involve the full Linux kernel. This is what I plan for [lxmonika](https://github.com/trungnt2910/lxmonika/discussions/5#discussioncomment-11131893). The approach might be difficult for Windows due to licensing (building a `lxcore.sys` based on Linux will get `winload.efi` infected by the GPL) and other issues, but is possible for an open-source project with enough time investments.

> But I have different view of this path, like WSX (Windows Subsystem for XNU) which will run some app like xcode like darling, which will contain some code from WinObjC and [ipasim](https://github.com/ipasimulator/ipasim).

In my opinion, library-level compatibility like what Darling and `ipasim` are doing will not be sustainable. Interfaces provided by user-mode libraries are seemingly infinite, they are much wider than that provided by the kernel (just a few hundred system calls and maybe a hundred more magic `ioctl`s. My vision of the Windows Subsystem for XNU would be something that uses the full XNU kernel as a Pico/`lxmonika` provider, then tries to run unmodified macOS libraries and frameworks.

> Are you Vietnamese? Can we share contact together?

If you're interested in contacting me, try joining the [Project Reality Discord](https://reality.trungnt2910.com/discord). We have a dedicated Vietnamese channel.

**Reply by VOM7HC** [reply]
Thank Trung, sorry make you mis understand,

>In my opinion, library-level compatibility like what Darling and ipasim are doing will not be sustainable. Interfaces provided by user-mode libraries are seemingly infinite, they are much wider than that provided by the kernel (just a few hundred system calls and maybe a hundred more magic ioctls. My vision of the Windows Subsystem for XNU would be something that uses the full XNU kernel as a Pico/lxmonika provider, then tries to run unmodified macOS libraries and frameworks.

WinObjC is port for ObjC runtime, mostly still for user land mode
for kernel mode, still base BSD and OSFMK

but that mean I show to you my view my idea

**Reply by kevincrans** [reply]
I was for fun making up something similar recently in the ravynOS community: running a macOS/Darwin BSD userland on top of the ReactOS NT kernel.
So maybe RSBSD is an idea, in the hands of the community.
I agree and think that WSL2 is not WSL, and WSL1 could have been faster if done right, & more compatible with a lot of effort.

**Reply by trungnt2910** [reply]
I think we already have most of the puzzle pieces for a WSL1-like approach for FreeBSD on NT 10.0:

1. A (fairly) stable way to create non-WSL Pico processes: https://github.com/trungnt2910/lxmonika
2. A way to statically embed an ELF shared object into a PE executable: https://github.com/trungnt2910/REPEAT
3. A way to produce a library containing the high-level layers of the target kernel with no hardware-specific dependencies. 
- For Linux, we can already harvest the objects from a UML compilation. 
- I am not sure if something similar is available for FreeBSD? 

All it takes now is $200 for ChatGPT to write some glue code and an engineer motivated enough to craft the prompts. 

Alternatively, I have heard of someone writing a fully-userland compatibility layer for OpenBSD, which proves that you may not need kernel intervention to emulate a Sys-V ABI on Windows.

**Comment by Coolis1362** [comment]
Don't Worry I Will Make WSBSD, Starting With FreeBSD Support At WSBSD

**Reply by TonyGonzalez0379** [reply]
Looking forward to seeing it

**Reply by Coolis1362** [reply]
Here's The Link: [https://github.com/Coolis1362/WSBSD](https://github.com/Coolis1362/WSBSD)

**Reply by Coolis1362** [reply]
Done with My First Release, 1.0.0.0, Here's The Link!: [https://github.com/Coolis1362/WSBSD/releases/tag/1.0.0.0](https://github.com/Coolis1362/WSBSD/releases/tag/1.0.0.0)

**Reply by Coolis1362** [reply]
Bruh I Am Getting Thumps Down From @trungnt2910 ?

**Reply by trungnt2910** [reply]
I think there is much more to building a subsystem than publishing a few `.bat` files and suspicious `.exe`s here and there.

**Reply by unrooted** [reply]
@Coolis1362 in all honesty, I appreciate progress on this so soon, however, I'd avoid making users to run some sketchy .exe files, at least provide instructions on how to build it ourselves 

other than that, I'm keeping fingers crossed for you, so enjoy a star added to your project.

**Reply by Coolis1362** [reply]
Dude It's Safe!

**Reply by Coolis1362** [reply]
BTW It's My 2nd Time Making A Subsystem! Just Don't Hate Me!

**Reply by unrooted** [reply]
no one's hating on you, in all honesty, it looks pretty cool

however, running random .exe files off of the internet is not the best approach in the long run, so you'd want to have this project fully open-sourced sooner or later

either that, or providing a paid app available via MS Store (can still be open-source, and buying it via MS Store would be a form of 'donation' to your project)

**Reply by trungnt2910** [reply]
By your own claims, you have more experience creating subsystems than myself.

So I expect your subsystem to be at least a little bit better than my [Monix](https://github.com/trungnt2910/lxmonika/tree/master/mxss), and can run some simple FreeBSD binaries.

It is not hate - it is just me pointing out either a huge misconception, severe ignorance, or maybe just a scam.

**Reply by xzenor** [reply]
> Dude It's Safe!

You saying it is, doesn't make it so. I know zero about writing a subsystem but I agree with everyone here on this and I'm gonna run a random executable. It could be amazing and work like a breeze but I'm not risking it.

**Reply by trungnt2910** [reply]
If you have the comprehension skills of a high school student and have read Microsoft's blogs on WSL1 and the so-called WSBSD repo, you will find out that they're just some poorly written batch files (probably vibe-coded) and no subsystem at all.

**Reply by Coolis1362** [reply]
I'm Using Visual studio i'm Just Having Trobles With The git GUI In VS code And Visual Studio.

**Reply by Coolis1362** [reply]
Ok I Did Used Soem AI To Help Me But Please Just Don't Hate Me

**Reply by Coolis1362** [reply]
hello, i know have 11 stars, 2 people watching, and thhese websites: https://coolis1362.github.io/wsbsd.org/ and https://ttsconsulting.github.io/openbsd.wsbsd.org/ (Don't ask me why i am a part of a org, thogh i am 1 year until my 10s)

**Reply by VOM7HC** [reply]
Sorry coolis, I reviewed your code not do anything here. With only managed code c# it is not subsystem.

**Reply by trungnt2910** [reply]
🤫 Don't break the kid's fantasy... He's doing quite well for a 9-year-old.

Everyone knows that as of now, the only known creators of Windows subsystems are Microsoft and [Project Reality](https://go.trungnt2910.com/monika) 👀

**Reply by VOM7HC** [reply]
Haha Trung, I don't extinguish the fire inside anyone, but some one just have background python script and their backend code is managed code with c# with support LLM and vibe coding style could not produce any real product here.

Microsoft has very long term with they interix and now with WSL, if one guy can make subsystem better them, he should pointed as CEO M$ instead monkey coder here 😁

**Reply by trungnt2910** [reply]
> managed code with c#

This is not an issue. NativeAOT is quite powerful and can produce valid PE files. With some support scripts, they can be signed and converted to a driver.

> with support LLM and vibe coding style

This is the real issue. Regardless of what other people say, AI is too stupid to do any real work, especially something that requires accuracy and strict compliance like implementing a subsystem.

> if one guy can make subsystem better them

I bet many people can. The average PhD candidate in Computer Science can probably code better than the current M$ CEO.
But M$ is an enterprise, not an academic institution. They are all about money and influence, not about advancing technology.

**Reply by VOM7HC** [reply]
> This is not an issue. NativeAOT is quite powerful and can produce valid PE files. With some support scripts, they can be signed and converted to a driver.

Haha, yeah you can produce nativeAOT or use unsafe classic c# or deprecated .net native to produce native code. But I just "provoked" him/her/them code because it only console print

> I bet many people can. The average PhD candidate in Computer Science can probably code better than the current M$ CEO.

I know dude, some one can work. But how they can have money from that. This is business choice. I just "provoked" him/her/them when him/her/them release WSBSD just 2 week after M$ open source it.

**Comment by pleasesendmeloadsofvegetables** [comment]
Once again, I'm asking for an official stance on this. As I mentioned [previously](https://github.com/microsoft/WSL/discussions/6540#discussioncomment-12294666), the last time a maintainer posted in this discussion was back in 2017. Especially now that WSL has been open-sourced, an official stance would be great.

**Comment by pgiffuni** [comment]
The WSL2 approach is being [worked on](https://x.com/balajesankar/status/1970585411153207715) (outside Microsoft):
https://github.com/BalajeS/WSL-For-FreeBSD

**Reply by Maseru898** [reply]
Seems like Development halted basically after you posted this :D

**Reply by ManaHime** [reply]
There seem to have been a talk at [BSDCan](https://x.com/facepalm_tar_gz/status/2070735741634982231?s=20)?

**Comment by Maseru898** [comment]
Any Update? Been 9 years now, i think enough time to make it work
